lib/shelf-reference.js

```diff
@@ -0,0 +1,116 @@
+"use strict";
+
+module.exports = (logger, host, requestPromise, ShelfFile, utils) => {
+    /**
+     * A reference to a Shelf bucket.
+     */
+    class ShelfReference {
+        /**
+         * @param {string} refName
+         * @param {string} authToken
+         */
+        constructor(refName, authToken) {
+            this.authToken = authToken;
+            this.refName = refName;
+        }
+
+
+        /**
+         * Creates and returns a ShelfFile at the given path.
+         *
+         * @param {string} path
+         * @param {Boolean} unique Denotes if the uploaded content should have
+         *                         a unique url.
+         * @return {ShelfFile}
+         */
+        getArtifact(path, unique) {
+            var uri;
+
+            logger.debug("In ShelfReference.getArtifact()");
+            uri = utils.buildUrl(path, host, this.refName, unique);
+
+            return new ShelfFile(path, host, uri, this.authToken);
+        }
+
+
+        /**
+         * Sends and returns a request with the provided search criteria.
+         *
+         * @param {string} pathToSearch The path to perform the search on.
+         * @param {string} search The search criteria.
+         * @param {(string|string[])} sort A string or array of strings for
+         *                                 how the results will be sorted.
+         * @return {Promise.<Object>}
+         */
+        search(pathToSearch, search, sort) {
+            var options, url;
+
+            logger.debug("In ShelfReference.search()");
+
+            url = utils.buildUrl(pathToSearch, host, this.refName);
+            options = this.getOptions();
+            options.url = `${url}/_search`;
+            options.body = {
```

Comment:
We should probably talk about this outside of this pull request so I can get more details as to everything I should validate.